js/spreadsheet.js

```diff
@@ -68,244 +68,196 @@ var spreadsheet = function(options) {
-          firstRow = i ? i - 1 : i;
-        }
-      }
+    return currentIndex;
+  }
 
-      firstRow = firstRow || 0;
+  /**
+   * Function to get coordinates which we need to select
+   * 
+   * @param {Array} cellPosition - an array of the selected coordinates
+   * 
+   * @returns {Array/Boolean} - coordinates that need to be selected. Example of the returned data: [[startRow, startCol, endRow, endCol]]
+   */
+  function coordinatesToSelect(cellPosition) {
+    var firstCol, lastCol, firstRow, lastRow, tableData;
+
+    // Returns array of the data from the table with Handsontable API
+    tableData = hot.getData();
+    selectedCellPosition = {
+      row: cellPosition[0][0],
+      col: cellPosition[0][1]
+    };
 
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][lastCol] === null) {
-          lastRow = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.top) {
-      // When data located on the top of the selected cell
-      lastRow = startAt[0];
+    var nearestDataAt = nearestCellsWithData(selectedCellPosition, tableData);
 
-      for (var i = lastRow - 1; i > 0; i--) {
-        if (allData[i][startAt[1]] === null) {
-          firstRow = i;
-          break;
-        }
-      }
+    // Return false in the case when we need to select all table data
+    if (nearestDataAt.selectAll) {
+      return false;
+    }
 
-      firstRow = firstRow || 0;
+    if (nearestDataAt.left) {
+      var col = selectedCellPosition.col - 1;
 
-      for (var i = startAt[1]; i > 0; i--) {
-        if (allData[firstRow][i] === null) {
-          firstCol = i ? i + 1 : i;
-          break;
-        }
-      }
+      firstCol = getStartIndex(tableData[selectedCellPosition.row], col);
 
-      firstCol = firstCol || 0;
+      // If column index is not 0 it means that data what we need to select is in the next column
+      firstCol = firstCol === 0 ? firstCol : firstCol + 1;
 
-      for (var i = firstCol; i < allData.length; i++) {
-        if (allData[firstRow][i] === null) {
-          lastCol = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.bottom) {
-      // When data located on the bottom of the selected cell
-      firstRow = startAt[0];
+      firstRow = getStartIndex(tableData, selectedCellPosition.row, firstCol);
 
-      for (var i = firstRow + 1; i < allData.length; i++) {
-        if (allData[i][startAt[1]] === null) {
-          lastRow = i - 1;
-          break;
-        }
-      }
+      // If row index is not 0 it means that data what we need to select is in the next row
+      firstRow = firstRow === 0 ? firstRow : firstRow + 1;
 
-      for (var i = startAt[1]; i > 0; i--) {
-        if (allData[lastRow][i] === null) {
-          firstCol = i + 1;
-          break;
-        }
-      }
+      lastCol = getEndIndex(tableData[firstRow], firstCol);
+      lastCol = lastCol < selectedCellPosition.col ? selectedCellPosition.col : lastCol;
```

Comment:
```suggestion
      lastCol = Math.min(lastCol, selectedCellPosition.col);
```

Reply:
@tonytlwu, in this case, we should use `Math.max` as well. Because `Math.min` will cause this issue https://share.getcloudapp.com/wbu0AeAb